Supporting debugging in extraction plugin by default

herbelin opened the pull request:
We add extraction directory and plugin to the call to ocamldebug.

Not sure all three incantations are necessary, but it allows printing (though setting breakpoints still fails, but this is an ocamldebug bug with packed modules, afaik).

SkySkimmer commented:
What's the reason for this?

SkySkimmer commented:
Ah `print` doesn't work without the -I from ocamlfind 
Maybe we should list all plugins there somehow?

I don't think we should add the cma, that's useless AFAIK

SkySkimmer commented:
>except setting breakpoints but this is an ocamldebug bug with packed modules, afaik

Setting breakpoints works if you remember that the module names have an automatic prefix (`Extraction_plugin__Extraction` instead of `Extraction`)
The coqdev.el automatically translates breakpoint requests in directories where the dune file has only 1 `(library` (and for files from the first `(library` otherwise).

herbelin commented:
> Ah print doesn't work without the -I from ocamlfind

Yes. I updated the header to clarify.

> I don't think we should add the cma, that's useless AFAIK

Right. I removed them then.

>  Maybe we should list all plugins there somehow?

Probably, do you know how to do it?

> Setting breakpoints works if you remember that the module names have an automatic prefix 

Could you give an example (in Emacs, C-x C-a C-b says "Can't find any event there")?

SkySkimmer commented:
>Could you give an example (in Emacs, C-x C-a C-b says "Can't find any event there")?

you need to have the plugin loaded first. So on something short like 
~~~coq
Require Extraction.

Extraction foo.
~~~
you can just hit `run` then after it finishes breakpoint and `reverse`
If there's a lot between the end of execution and where you want to break, either you `run` with `set break_on_load on` until it's loaded, or you need to find a convenient breakpoint between the loading and the command you want to work on
for instance
~~~coq
Require Extraction.

(* some commands *)

Comments is a nice trick. (* set initial breakpoint near https://github.com/coq/coq/blob/309a71cd105eee9c6f2ba380fe6f6005ff7afeb8/vernac/vernacentries.ml#L2642 *)

(* some commands *)

Extraction foo.
~~~

But TBH when I want to debug a plugin I tend to statically link it (put it in the list at https://github.com/coq/coq/blob/309a71cd105eee9c6f2ba380fe6f6005ff7afeb8/topbin/dune#L26) so it just works.
It seems to make `print` work too.

herbelin commented:
> Comments is a nice trick. 

What I do generally, is to put a breakpoint after the `with exn ->` in `Vernacinterp.interp_gen` so that any failing command returns back to the debugger.

> But TBH when I want to debug a plugin I tend to statically link it (put it in the list at ...

Thanks for telling where to put the information (I knew for make but not for dune).

You answer seems otherwise to suggest that you don't use C-x C-a C-b either?

SkySkimmer commented:
I do use C-x C-a C-b
It works thanks to https://github.com/coq/coq/blob/309a71cd105eee9c6f2ba380fe6f6005ff7afeb8/dev/tools/coqdev.el#L171

herbelin commented:
Thanks a lot! It is that my `coqdev.el` was outdated.

SkySkimmer commented on dev/dune-dbg.in:
```diff
@@ -45,4 +45,4 @@ while [[ $# -gt 0 ]]; do
 done
 
-ocamldebug "${opts[@]}" $(ocamlfind query -recursive -i-format coq-core.top_printers) -I +threads -I dev $exe "$@"
+ocamldebug "${opts[@]}" $(ocamlfind query -recursive -i-format coq-core.top_printers coq-core.plugins.extraction) -I +threads -I dev $exe "$@"
```
```suggestion
ocamldebug "${opts[@]}" $(ocamlfind query -recursive -i-format coq-core.top_printers) $(ocamlfind query -i-format -descendants coq-core.vernac) -I +threads -I dev $exe "$@"
```
this seems to work to get all plugins

herbelin replied:
Works at least for extraction. Committed. Thanks.

SkySkimmer commented:
@coqbot merge now
